cmd/soroban-rpc/internal/db/event.go

```diff
@@ -97,3 +109,156 @@ func (eventHandler *eventHandler) InsertEvents(lcm xdr.LedgerCloseMeta) error {
+	if latestLedgerSeq+1 <= retentionWindow {
+		return nil
+	}
+
+	cutoff := latestLedgerSeq + 1 - retentionWindow
+	_, err := sq.StatementBuilder.
+		RunWith(eventHandler.stmtCache).
+		Delete(eventTableName).
+		Where(sq.Lt{"ledger_sequence": cutoff}).
+		Exec()
+	return err
+}
+
+// GetEvents applies f on all the events occurring in the given range with specified contract IDs if provided.
+// The events are returned in sorted ascending Cursor order.
+// If f returns false, the scan terminates early (f will not be applied on
+// remaining events in the range).
+func (eventHandler *eventHandler) GetEvents(
+	ctx context.Context,
+	cursorRange CursorRange,
+	contractIDs [][]byte,
+	f ScanFunction,
+) error {
+	start := time.Now()
+
+	var rows []struct {
+		TxIndex int                 `db:"application_order"`
+		Lcm     xdr.LedgerCloseMeta `db:"meta"`
+	}
+
+	rowQ := sq.
+		Select("e.application_order", "lcm.meta").
+		From(eventTableName + " e").
+		Join(ledgerCloseMetaTableName + " lcm ON (e.ledger_sequence = lcm.sequence)").
+		Where(sq.GtOrEq{"e.ledger_sequence": cursorRange.Start.Ledger}).
+		Where(sq.GtOrEq{"e.application_order": cursorRange.Start.Tx}).
+		Where(sq.Lt{"e.ledger_sequence": cursorRange.End.Ledger}).
+		OrderBy("e.ledger_sequence ASC")
+
+	if len(contractIDs) > 0 {
+		rowQ = rowQ.Where(sq.Eq{"e.contract_id": contractIDs})
+	}
+
+	if err := eventHandler.db.Select(ctx, &rows, rowQ); err != nil {
+		return fmt.Errorf(
+			"db read failed for start ledger cursor= %v contractIDs= %v: %w",
+			cursorRange.Start.String(),
+			contractIDs,
+			err)
+	} else if len(rows) < 1 {
+		eventHandler.log.Debugf(
+			"No events found for ledger range: start ledger cursor= %v - end ledger cursor= %v contractIDs= %v",
+			cursorRange.Start.String(),
+			cursorRange.End.String(),
+			contractIDs,
+		)
+		return nil
+	}
+
+	for _, row := range rows {
+		txIndex, lcm := row.TxIndex, row.Lcm
+		reader, err := ingest.NewLedgerTransactionReaderFromLedgerCloseMeta(eventHandler.passphrase, lcm)
+		if err != nil {
+			return fmt.Errorf("failed to create ledger reader from LCM: %w", err)
+		}
+
+		err = reader.Seek(txIndex - 1)
+		if err != nil {
+			return fmt.Errorf("failed to index to tx %d in ledger %d: %w", txIndex, lcm.LedgerSequence(), err)
+		}
+
+		ledgerCloseTime := lcm.LedgerCloseTime()
+		ledgerTx, err := reader.Read()
+		if err != nil {
+			return fmt.Errorf("failed reading tx: %w", err)
+		}
+		transactionHash := ledgerTx.Result.TransactionHash
+		diagEvents, diagErr := ledgerTx.GetDiagnosticEvents()
+
+		if diagErr != nil {
+			return fmt.Errorf("couldn't encode transaction DiagnosticEvents: %w", err)
+		}
+
+		// Find events based on filter passed in function f
+		for eventIndex, event := range diagEvents {
+			cur := Cursor{Ledger: lcm.LedgerSequence(), Tx: uint32(txIndex), Event: uint32(eventIndex)}
+			if !f(event, cur, ledgerCloseTime, &transactionHash) {
+				return nil
+			}
+		}
+	}
```

Comment:
Could we parallelize this, maybe?? Unless the speed is acceptable - goroutines would probably increase allocs

Reply:
Latency has improved significantly with the new design so I have not experimented with it but we will refactor as we go forward. but honestly performance should not degrade overall for quite some time as current traffic is very low.